src/engine/fileOpsIns.cpp

```diff
@@ -634,6 +637,200 @@ void DivEngine::loadSBI(SafeReader& reader, std::vector<DivInstrument*>& ret, St
+  DivInstrument* ins = new DivInstrument;
+
+  try {
+    reader.seek(0, SEEK_SET);
+
+    String header = reader.readString(11);
+    if (header.compare("WOPN2-INST") == 0 || header.compare("WOPN2-IN2T") == 0) {  // omfg
+      uint16_t version = reader.readS();
+      if (!(version >= 2) || version > 0xF) {
+        // version 1 doesn't have a version field........
+        reader.seek(-2, SEEK_CUR);
+      }
+
+      reader.readC(); // skip isPerc
+      ins->type = DIV_INS_FM;
+      ins->fm.ops = 4;
+
+      String insName = reader.readString(32);
+      ins->name = (insName.length() > 0) ? insName : stripPath;
+      reader.seek(3, SEEK_CUR);  // skip MIDI params
+      uint8_t feedAlgo = reader.readC();
+      ins->fm.alg = (feedAlgo & 0x7);
+      ins->fm.fb = ((feedAlgo>>3) & 0x7);
+      reader.readC();  // Skip global bank flags - see WOPN/OPNI spec
+
+      auto readOpniOp = [](SafeReader& reader, DivInstrumentFM::Operator& op) {
+        uint8_t dtMul = reader.readC();
+        uint8_t totalLevel = reader.readC();
+        uint8_t arRateScale = reader.readC();
+        uint8_t drAmpEnable = reader.readC();
+        uint8_t d2r = reader.readC();
+        uint8_t susRelease = reader.readC();
+        uint8_t ssgEg = reader.readC();
+
+        op.mult = dtMul & 0xF;
+        op.dt = ((dtMul >> 4) & 0x7);
+        op.tl = totalLevel & 0x3F;
+        op.rs = ((arRateScale >> 6) & 0x3);
+        op.ar = arRateScale & 0x1F;
+        op.dr = drAmpEnable & 0x1F;
+        op.am = ((drAmpEnable >> 7) & 0x1);
+        op.d2r = d2r & 0x1F;
+        op.rr = susRelease & 0xF;
+        op.sl = ((susRelease >> 4) & 0xF);
+        op.ssgEnv = ssgEg;
+      };
+
+      for (int i : {0,1,2,3}) {
+        readOpniOp(reader, ins->fm.op[i]);
+      }
+
+      // Skip rest of file
+      reader.seek(0, SEEK_END);
+      ret.push_back(ins);
+    }
+  } catch (EndOfFileException& e) {
+    lastError = "premature end of file";
+    logE("premature end of file");
+    delete ins;
+  }
+}
+
+void DivEngine::loadY12(SafeReader& reader, std::vector<DivInstrument*>& ret, String& stripPath) {  
+  DivInstrument *ins = new DivInstrument;
+
+  try {
+    reader.seek(0, SEEK_SET);
+    ins->type = DIV_INS_FM;
+    ins->fm.ops = 4;
+    ins->name = stripPath;
+
+    for (int i : {0,1,2,3}) {
+      DivInstrumentFM::Operator& insOp = ins->fm.op[i];
+      uint8_t tmp = reader.readC();
+      insOp.mult = tmp & 0xF;
+      insOp.dt = ((tmp >> 4) & 0x7);
+      insOp.tl = (reader.readC() & 0x3F);
+      tmp = reader.readC();
+      insOp.rs = ((tmp >> 6) & 0x3);
+      insOp.ar = tmp & 0x1F;
+      tmp = reader.readC();
+      insOp.dr = tmp & 0x1F;
+      insOp.am = ((tmp >> 7) & 0x1);
+      insOp.d2r = (reader.readC() & 0x1F);
+      tmp = reader.readC();
+      insOp.rr = tmp & 0xF;
+      insOp.sl = ((tmp >> 4) & 0xF);
+      insOp.ssgEnv = reader.readC();
+      reader.seek(9, SEEK_CUR);
```

Comment:
Check return value.

Reply:
No need as exception thrown will handle post-EOF case.

Reply:
Decided you are correct and added exception throw on false.